### PR TITLE
Old mmap->lazyio, add new mmap

### PR DESCRIPTION
**Pull request by timholy**
This PR reworks the handling of "mmapped" images, preserving the
previous implementation and making it accessible with `lazyio=true`.
The main motivation is to take advantage of the OS-provided buffering
of mmapped data, so that elementwise access patterns become much
more efficient, particularly for 3d images. (In some simple demos,
by multiple orders of magnitude.)

I went with a "name swap" because calling the previous implementation
"mmap" is a misnomer: it is a useful form of lazy-loading, but
there was no use of "true" mmapping in the sense of `Mmap.mmap`.

This also introduces the ability to open a mmapped TIFF with `r+`
permissions and thus write (simply by setting values) as well as read.

Here's a quick demo of the performance improvement:

```julia
julia> using TiffImages, ImageCore

julia> img0 = rand(Gray{N0f16}, 1000, 1000, 40);

julia> TiffImages.save("/tmp/stack.tif", img0);

julia> img1 = TiffImages.load("/tmp/stack.tif");

julia> img2 = TiffImages.load("/tmp/stack.tif"; mmap=true);

julia> img3 = TiffImages.load("/tmp/stack.tif"; lazyio=true);

julia> using BenchmarkTools

julia> @btime $img1[1,1,1] + $img1[1,1,2]
  2.839 ns (0 allocations: 0 bytes)
Gray{N0f16}(0.01758)

julia> @btime $img2[1,1,1] + $img2[1,1,2]
  7.432 ns (0 allocations: 0 bytes)
Gray{N0f16}(0.01758)

julia> @btime $img3[1,1,1] + $img3[1,1,2]
  231.848 μs (21 allocations: 496 bytes)
Gray{N0f16}(0.01758)
```

In this particular case it's an improvement of more than 4 orders of magnitude. The reason is that by switching slice planes, the `lazyio=true` version has to re-read the active slice each time. 

What this should enable is applications like https://imagej.net/plugins/bdv/ in which the slice visualization plane may be chosen arbitrarily, without forcing users to write their data in 3d chunks using HDF5 or similar formats.

Closes https://github.com/JuliaArrays/MappedArrays.jl/issues/49

### CODE REVIEW

**Comment by timholy** [comment]
I was going to use `lazy=true`, but it turns out that [ImageIO now uses that](https://github.com/JuliaIO/ImageIO.jl/pull/49) in the sense of "lazy modules" and here I want it to mean "lazy data access".

**Review by tlnagy — CHANGES_REQUESTED** [review]
This looks great! Would be helpful to update the docs?

**Comment by tlnagy** [comment]
Thanks @timholy. I think this is a reasonable set of changes. Less flexible, but substantially better worse case behavior. 

The original approach is a disaster if you do things like https://github.com/JuliaArrays/MappedArrays.jl/issues/49, this PR basically solves that problem completely.

Could you update the precompile statement to improve the time to first image?
https://github.com/tlnagy/TiffImages.jl/blob/facdf3dc931e0974fdde291b1002b39ee0a29c06/src/TiffImages.jl#L39-L46

**Comment by timholy** [comment]
Also, something worth considering: I suspect that long-term, we want `mmap=true` to fail rather than fall back to `lazyio=true`. The reason is that the performance differences are dramatic, and people may think that they're using `mmap` when instead they're actually using `lazyio`.

One *could* call this effectively a deprecation warning, only bumping the minor version when `mmap=true` becomes a hard-fail for unsupported file types. However, I suspect that's risky, and it may be better to call this 0.6 and then think about removing the fallback for `lazyio` in v0.7 (or whatever you want to call it).

File: src/TiffImages.jl
Comment (johnnychen94) [inline src/TiffImages.jl:44]: This works locally

```suggestion
            TiffImages.save(fpath, rand(c{t}, sz))
            Sys.iswindows() && GC.gc()
```

Comment (timholy) [reply src/TiffImages.jl:44]: Huh, weird. Do you understand why? It's the loading via `mmap` that's presumably causing the problem, so it's not clear to me why GCing after the save helps. Maybe it's just an extra GC on each loop that does a more thorough cleaning?

**Comment by johnnychen94** [comment]
Yeeeah~ :tada: some new errors 😂

**Comment by johnnychen94** [comment]
I have access to a Windows machine, let me try to fix it locally first.

File: src/types/mmapped.jl
Comment (johnnychen94) [inline src/types/mmapped.jl:31]: > I have access to a Windows machine, let me try to fix it locally first.

I believe https://github.com/timholy/TiffImages.jl/pull/1 fixes the test. But it seems to uncover a windows caveat that mmap-ed images should be surrounded by a `let` block and eager `GC.gc()`

https://github.com/timholy/TiffImages.jl/pull/2 is an attempt to eagerly do `GC.gc()` if issues like this happens. But one should still do `GC.gc()` before trying to remove the mmapped files.

Comment (johnnychen94) [reply src/types/mmapped.jl:31]: MWE:

```julia
using ImageCore
using TiffImages
using ImageShow

img = rand(Gray, 64, 64)
TiffImages.save("tmp.tif", img)

img = TiffImages.load("tmp.tif"; mmap=true)

img2 = TiffImages.load("tmp.tif") # works
TiffImages.save("tmp.tif", img2) # fails -- because img occupies the file handler
```

With https://github.com/timholy/TiffImages.jl/pull/2

```julia
img2 = let img = TiffImages.load("tmp.tif"; mmap=true)
    TiffImages.load("tmp.tif") # works
end
TiffImages.save("tmp.tif", img2) # works -- with an warning
```

```julia
┌ Warning: failed to open file tmp.tif in "w" mode, retry after GC.gc()
└ @ TiffImages d:\Julia\TiffImages.jl\src\utils.jl:205
```

Comment (johnnychen94) [reply src/types/mmapped.jl:31]: Maybe we should not use the `load` interface, but instead, use something like the following since mmap requests a resource so it should follow the RAII (Resource Acquisition Is Initialization) practice.

```julia
TiffImages.lazy_load(filepath) do img
    ...
end
```

**Comment by timholy** [comment]
Let's wait for https://github.com/timholy/TiffImages.jl/pull/2 before merging this.

File: src/load.jl
Comment (johnnychen94) [inline src/load.jl:29]: > Let's wait for https://github.com/timholy/TiffImages.jl/pull/2 before merging this.

I now don't believe #2 is the correct fix. We should fix the semantics instead. 

Note the `close(tf.io)` in R42 -- this means `mmap=true` never correctly close the file IO -- the fact that linux CI passes the test is just some bad luck.

Instead, we should use the explicit `mmap` semantic in https://github.com/timholy/TiffImages.jl/pull/3 and it's a perfect fix for https://github.com/tlnagy/TiffImages.jl/pull/79/files#r880478304

```julia
img = rand(Gray, 64, 64)
TiffImages.save("tmp.tif", img)
sz = TiffImages.mmap_load("tmp.tif") do img
    size(img)
end
img2 = TiffImages.load("tmp.tif") # works
TiffImages.save("tmp.tif", img2) # works
```

---

We could even bring this `mmap` function to FileIO, although I'm not sure if it is feasible.

Comment (timholy) [reply src/load.jl:29]: >  the fact that linux CI passes the test is just some bad luck.

Not really, see https://stackoverflow.com/questions/42961339/mmap-after-deleting-the-file. I wish Windows did this. I think that's really the root of all the problems on Windows, and everything works nicely as you'd hope on Linux.

**Comment by timholy** [comment]
I think these problems with Windows files occur mostly in tests. How often do you really expect this to happen in practice? If someone has gone to the effort to collect a 100GB image series, are they really likely to just delete it?

File: examples/mmap_lazyio.jl
Comment (johnnychen94) [inline examples/mmap_lazyio.jl:46]: Otherwise, this will a plain markdown line with no syntax highlight.

img = TiffImages.load(filepath; mmap=true, mode="r+")

````suggestion
# ```julia
# img = TiffImages.load(filepath; mmap=true, mode="r+")
# ```
````

File: src/types/mmapped.jl
Comment (johnnychen94) [inline src/types/mmapped.jl:76]:
```julia
julia> img = TiffImages.load("tmp.tif"; mmap=true)

julia> img[2] = img[1]
ERROR: ReadOnlyMemoryError()
Stacktrace:
  [1] setindex!
    @ .\array.jl:903 [inlined]
  [2] setindex!
```

An idea here can be explicitly constructing the `MmappedTIFF` with a readonly bool value so that we can eagerly detect and throw errors.

```julia
struct MmappedTIFF{T <: Colorant, N, Writable, O <: Unsigned, A <: AbstractMatrix{T}} <: AbstractTIFF{T, N}
...
end
...

Base.@propagate_inbounds Base.setindex!(img::MmappedTIFF{T, 2, true}, val, i::Int, j::Int) where T = chunk1(img)[j, i] = val
Base.@propagate_inbounds function Base.setindex!(img::MmappedTIFF{T, 3}, val, i::Int, j::Int, k::Int) where T
    chunk = img.chunks[k]
    chunk[j, i] = val
end
function Base.setindex!(img::MmappedTIFF{T, N, false}, val, i::Int, j::Int) where {T,N}
    error("unable to write a readonly memory-mapped array, maybe you need to open the file with `r+` mode.")
end
```

File: examples/mmap_lazyio.jl
Comment (johnnychen94) [inline examples/mmap_lazyio.jl:52]: > I think these problems with Windows files occur mostly in tests. How often do you really expect this to happen in practice? If someone has gone to the effort to collect a 100GB image series, are they really likely to just delete it?

Although I believe the right fix is to not abuse the `load` interface, I get your point that we probably
don't need to bother with this edge case issue. Just that I'm not sure how broadly this affects people's
daily life usage.

But I would suggest a clear warning on this issue, especially for windows, and I believe we should also
explicitly document the behavior that the opened file handler will be delayed to Julia's GC somewhere
in the docstring.

**Review by tlnagy — APPROVED** [review]

**Comment by tlnagy** [comment]
This PR looks good from my end. I'm open to whatever you all think re: Windows. I won't merge until I get the all clear from @timholy. It's probably a good idea to withhold from updating `ImageIO.jl` for a bit to let bug reports to come in once I tagged version 0.6.0.

File: src/load.jl
Comment (johnnychen94) [inline src/load.jl:31]: IIUC, this is a file-specific warning so should be enabled for every file, not just the first one.

```suggestion
        mmap && @warn "Compression and discontiguous planes are not supported by `mmap`, use `lazyio = true` instead"
```

File: src/types/mmapped.jl
Comment (timholy) [reply src/types/mmapped.jl:76]: That's a good idea. Just to decrease need for compilation specialization (some may use both read-only and read-write in the same session), do you think we could do almost as well with `register_error_hint`? If so, that would require less duplication of compilation effort.

Comment (johnnychen94) [reply src/types/mmapped.jl:76]: The duplication of compilation is a valid reason to not do this.

But I'm not sure whether we can register an error hint for `ReadOnlyMemoryError` because [it's an empty struct](https://github.com/JuliaLang/julia/blob/6ce28008ba6db324b171909fa8e641fe8bce9db4/base/boot.jl#L282-L287) that we have no extra information
to examine and filter. Can `register_error_hint` get callstack information in general?

If not doable, the current best approach is just to mention `ReadOnlyMemoryError` error is due to file mode in the documentation.

**Review by johnnychen94 — APPROVED** [review]
Very informative documentation! I guess we just need to merge https://github.com/timholy/TiffImages.jl/pull/2 before merging this PR?

**Comment by timholy** [comment]
Thanks!This was a very productive discussion and I think users will be better-served for it.

> I guess we just need to merge https://github.com/timholy/TiffImages.jl/pull/2 before merging this PR?

I think that makes sense. I left a suggestion for an alternative implementation, but we can go with yours if you prefer it.

> It's probably a good idea to withhold from updating ImageIO.jl for a bit to let bug reports to come in once I tagged version 0.6.0.

We should also suppress "canonicalization" if either `mmap` or `lazyio` keywords are present, see https://github.com/JuliaIO/ImageIO.jl/pull/51.

**Comment by timholy** [comment]
Presuming it passes tests, I think this is now ready to merge.

**Comment by tlnagy** [comment]
Given that we renamed mmap to lazyio, @timholy would you mind renaming the `DiskTaggedImage` type as well? The old name doesn't make sense since the new mmapped type is called `MmappedTIFF`.

**Comment by timholy** [comment]
How is `DiskTaggedImage` wrong? I'm not saying there isn't a better name, but I don't see anything misleading about it.

But I did the rename and went with `LazyBufferedTiff`.

**Comment by tlnagy** [comment]
It's not incorrect per se, but in light of the introduction of the `MmappedTIFF` type with this PR, it doesn't make much sense to keep the name `DiskTaggedImage` since they are both "on disk" TIFFs. `LazyBufferedTIFF` is better since it sticks with the lazy vs mmapped distinction we're going with.

**Comment by timholy** [comment]
Sounds good to me. I don't have merge privileges, so from my standpoint merge whenever you feel comfortable.

**Comment by tlnagy** [comment]
Thanks @timholy :tada: , this will be a welcome performance improvement!

**Pull request merged**
